### DISCUSSION TITLE
May NyPullToRefresh could be better

### DISCUSSION THREAD

**Discussion by jitendravn** [nylo-core/nylo#121] [General] [answered]
## My Data
{
        data: []
        pagination: {total: 15, count: 0, per_page: 10, current_page: 3, last_page: 2, hasMorePages: false}
 }
 
## Your widget error that i occured

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Null check operator used on a null value
E/flutter (24138): #0      _NyPullToRefreshState._onLoading (package:nylo_support/widgets/ny_pull_to_refresh.dart:161:16)

## and here is code your where the error occured

```Dart
_onLoading() async {
    _iteration++;
    List<T>? newData = await widget.data(_iteration);
    if (newData == null) _refreshController.loadNoData();
    if (newData!.isEmpty) {
      _refreshController.loadNoData();
      return;
    }

    _data.addAll(newData);

    if (mounted) {
      setState(() {});
    }
    _refreshController.loadComplete();
  }
  ```
  ## May be you could check for runtime type  then
  check for nullable to is empty or length
  so may be could be better for those kind data 
  i have pasted above
  

here is and i get while using ny_pull_to_refresh

**Comment by agordn52** [comment]
Hi @jitendravn,

Can you share your complete code?
I'm struggling to see the error.

Thanks

**Comment by jitendravn** [comment]
I have received these data from the server
 
 ```
 {
data: []
pagination: {total: 15, count: 0, per_page: 10, current_page: 3, last_page: 2, hasMorePages: false}
}
```
and I'm using  your PullToRefresh Widget 

![Screenshot 2024-04-19 at 11 49 42 AM](https://github.com/nylo-core/nylo/assets/93908273/ba16876f-4ba0-4a3e-8d87-25fc992ccead)

please see your ny_pull_to_refresh.dart 
file 
line not exactly but may its 
156

```Dart

  _onLoading() async {
    _iteration++;

    List<T>? newData = await widget.data(_iteration);
    if (newData == null) _refreshController.loadNoData();
    if (newData!.isEmpty) {
      _refreshController.loadNoData();
      return;
    }

    _data.addAll(newData);

    if (mounted) {
      setState(() {});
    }
    _refreshController.loadComplete();
  }

```

**Reply by agordn52** [reply, marked as answer]
Hi @jitendravn,

The `NyPullToRefresh` widget expects a return type of `List` in the `data` parameter.
You should modify your code to this.

``` dart 
data: (int iteration) async {
   YourDataType userCall = await api<HomePageApiService>((request) async => request.userCalllogs(iteration));
   return userCall.data;
},
```

Hope that helps.
